### PR TITLE
Run windows tests in multiple shells

### PR DESCRIPTION
**Pull request by blairconrad**
From https://github.com/pawamoy/duty/issues/23, which identified problems running failprint on Windows, specifically when _not_ run under a bash-like shell.

The current CI builds run the Windows tests using git bash and so don't surface the problem.

### CODE REVIEW

**Comment by pawamoy** [comment]
Great, so it passes with powershell. Could you now *remove* the fix (setting the env var) to see if it actually fails?

**Comment by pawamoy** [comment]
Hmmm actually it looks like the `shell` variable of the matrix is not used anywhere.

**Comment by pawamoy** [comment]
OK it's *not* implicit, because as you can see both the bash and pwsh jobs use `pwsh` to run the `make test` command. You'll probably have to use the matrix `shell` variable like this: https://docs.github.com/en/actions/writing-workflows/workflow-syntax-for-github-actions#jobsjob_idstepsshell.

**Comment by pawamoy** [comment]
Thank you for all this :rocket:

**Comment by blairconrad** [comment]
Ha, so many comments. Yes, I am an inexpert GitHub action person. I intended to flair about privately before embarrassing myself, but now I'm in the public eye. Expect way too many updates here. You might want to mute the thread!

**Comment by blairconrad** [comment]
I actually slightly fear that running `make` might pollute the results, as it could be forcing us into a "*nixy" situation. But that's just wild speculation at this point.

**Comment by pawamoy** [comment]
Valid concern :thinking: Well, we'll know for sure if you manage to get one test job to fail on PowerShell. If you can't make it fail, you might want to customize the step indeed, to run Python directly, without `make`.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -108,3 +103,45 @@ jobs:
...
+        cache-dependency-glob: pyproject.toml
+        cache-suffix: py${{ matrix.python-version }}
+
+    - name: Install dependencies
+      env:
+        UV_RESOLUTION: ${{ matrix.resolution }}
+      run: make setup
+
+    - name: Run the test suite
+      run: uv run duty --capture=none test
```
Comment (pawamoy) [inline .github/workflows/ci.yml:147]: Note that `--capture=none` won't trigger the issue as capture is completely disabled and lets both stdout and stderr flow into the upper console.

Comment (blairconrad) [reply .github/workflows/ci.yml:147]: Ah, but I'm keeping outer duty from capturing, not (I believe) inner failprint from capturing!

No?

Comment (pawamoy) [reply .github/workflows/ci.yml:147]: Ha, I think you're right! Sorry.

**Comment by pawamoy** [comment]
Flushing issue. You can try to add `flush=True` to the `print` call. For `sys.stdout.write`, I don't remember if there's such a `flush` parameter.

File: src/failprint/capture.py [hunk 2]
```diff
@@ -113,8 +113,15 @@ def __init__(self, capture: Capture = Capture.BOTH, stdin: str | None = None) ->
...
         self._saved_stderr_fd: int = -1
         self._output: str | None = None
+        self._old_env: dict[str, str | None] = {}
 
     def __enter__(self) -> CaptureManager:  # noqa: PYI034 (false-positive)
+        # Annoying operating system being annoying again.
+        # Bash ports and WSL2 work fine, this environment variable
+        # just helps a bit with cmd and PowerShell.
+        self._old_env["PYTHONLEGACYWINDOWSSTDIO"] = os.getenv("PYTHONLEGACYWINDOWSSTDIO")
+        os.environ["PYTHONLEGACYWINDOWSSTDIO"] = "1"
```
Comment (pawamoy) [inline src/failprint/capture.py:123]: Kay, testing without the fix now.

```suggestion
        #self._old_env["PYTHONLEGACYWINDOWSSTDIO"] = os.getenv("PYTHONLEGACYWINDOWSSTDIO")
        #os.environ["PYTHONLEGACYWINDOWSSTDIO"] = "1"
```

**Comment by pawamoy** [comment]
Too bad, tests passing even without the fix.

**Comment by blairconrad** [comment]
Hey. Work was busy. Sorry you had to turn off the fix. I'm flummoxed by the results I see. With the fix, I still have a poor user experience. My hand-rolled "smoketest" fails.

With this code:


```python
import failprint
import sys
import time

def passing():
    time.sleep(1)
    print("I'm passing")
    return 0

def failing():
    time.sleep(1)
    print("I'm failing")
    return 1

failprint.run(vars()[sys.argv[1]], capture=sys.argv[2])
```

* `uv run smoketest.py passing both` exits with code 120 and no output
* `uv run smoketest.py failing both` exits with code 120 and no output
* `uv run smoketest.py passing none` exits with code 0 and shows good output (including fancy glyphs)
* `uv run smoketest.py failing none` exits with code 0 and shows good output (including fancy glyphs)

If I set PYTHONLEGACYWINDOWSSTDIO="1" _first_, then all combinations above exit with code 1 and prints error indicating that "UnicodeEncodeError: 'charmap' codec can't encode character '\u2717' in position 5: character maps to <undefined>" (or `\u2713`)

If I then also set PYTHONIOENCODING="UTF-8" (as the github actions have) and repeat, the smoketests behave as you'd hope ("I'm failing" always printed, "I'm passing" only with capture=none), but the fancy glyphs are mangled, with what's probably their 3 UTF-8 bytes rendered directly as CP1252 characters (I'm too lazy to look, but "Γ£ô" and "Γ£ù". Feels right.)

I think running in pytest or possibly on github machines is confounding the results. I'm going to amend the job to also run the smoketests and then walk away. I may or may not check the result before bed. I'll keep your latest commit (undoing your fix) anyhow.

**Comment by pawamoy** [comment]
> but the fancy glyphs are mangled

Oh OK, I thought the result was "worse" than that. I'd consider it solved on failprint's side with `PYTHONLEGACYWINDOWSSTDIO=1`, and on the user side with `PYTHONUTF8=1` or `PYTHONIOENCODING=UTF8`. You might want to look deeper into this unicode encode/decode error on Windows to fix the glyphs display :shrug:

**Comment by blairconrad** [comment]
I did not expect a reply at 03:00! But anyhow, we currently seem to be at a point where user-side, we can workaround by setting PYTHONLEGACYWINDOWSSTDIO and maybe PYTHONIOENCODING (or just the former and using a different template). But so far failprint changes don't seem to be doing much good for your average Windows user...

**Comment by pawamoy** [comment]
🤫

I'll probably add the env var to failprint anyway  and add a small paragraph in the docs of both failprint and duty to recommend setting the UTF8 env vars for Windows users.

**Comment by blairconrad** [comment]
Sorry, I am a little lost. The environment variable, when set from within failprint, does nothing, as far as I can tell. I guess it would affect further pythons spawned from within failprint. But is there a real advantage to setting it within failprint?

(I also wonder if the naive trick of swapping out `sys.stdout` and `sys.stderr` would help. Coupled with duplicating the file descriptors, it might mean that both suprocess comamnds and callable commands that don't further spawn a subprocess would work... Of course this may very well be getting to be too much work to support an OS that you don't even use. And that is so far impossible to test on GitHub Actions.)

I am on the side investigating the cp1252 issue, but work interferes. Also, there's no stacktrace. You go from print to the complaint about the encoding!

**Comment by pawamoy** [comment]
> The environment variable, when set from within failprint, does nothing, as far as I can tell.

My understanding was that this would solve the "invalid handle" error, but since I can't try for myself, I can't say for sure. It *does* fix it for you, though you probably set the env var higher in the stack. I thought setting it right before doing the `os.dup2` hackery would have the same effect, i.e. fixing the "invalid handle" issue.

What our tests here have shown is not that setting the env var in failprint does nothing, but that we can't reproduce the "invalid handle" issue in GitHub Actions.

If you can confirm this "invalid handle" issue is indeed fixed by failprint setting the env var (and you *not* setting it yourself), I'll push and release this change. If it doesn't fix the "invalid handle" issue for you, there's indeed no point in making the change, and I'll only document the env var in a troubleshooting section.

Let me know if anything is unclear!

**Comment by blairconrad** [comment]
More on the environment variable. I never read the python source, but for CPython it seems to me the environment variable is read to establish initial configuration, right at startup, and then a config object is passed around to affect later behaviour:

https://github.com/python/cpython/blob/e545ead66ce725aae6fb0ad5d733abe806c19750/Python/initconfig.c#L1767-L1770

which would support the scenario where setting the environment variable after startup made no difference

**Comment by pawamoy** [comment]
I see, thank you for confirming! Then yeah, no need to change anything in failprint, and I'll just document this in a troubleshooting section :slightly_smiling_face:

**Comment by blairconrad** [comment]
Thanks. And I'm sorry it hasn't worked out better. I had high hopes, but it seems you're doing things that Windows just doesn't want you to.

Me, for now I'm probably going to try using duty with subprocess-based duties instead of callable. It's not satisfying, but will probably work well enough that I'll not notice any problems.

Out of curiosity, are you interested in either

1. adjustments to make it possible to run the failprint tests on Windows? I have been sitting on some workarounds for missing executables (test, fail, true, echo) that the GitHub Actions environment just seems to supply. I have a commit or so that shouldn't interfere with your current setup and would make for a better experience for Windows-based potential contributors
2. adjustments to your uv-copier template along the same lines? I was slowed down because I don't have make, and updates to the instructions might ease someone else's way. Also, a change to (e.g.) the `vscode` duty to not rely on `cp`.

I realize Windows isn't your target platform so if you'd rather not even see PRs, I will drop the idea right now. Otherwise, I'm happy to submit them for you to reject!

Again, sorry nothing more useful has come from the bug reports/PRs. Regardless, I've had a good time trying. You've been very accommodating.

**Comment by pawamoy** [comment]
1. Sure, we have a specific cross-platform use-case in failprint, so I'm OK to make it easier to test on all supported systems.
2. If these are small adjustments, why not. `make` is not actually required, only Python: you can run `python scripts/make` to obtain the command/task interface. So, OK for getting rid of `cp` in the `vscode` task. Other adjustments would have to be discussed :slightly_smiling_face:

**Comment by pawamoy** [comment]
And no worries, I enjoyed our interactions too and learned a few things, so all good!

**Comment by blairconrad** [comment]
Closing. As discovered, the GitHub Actions prevent us from testing with proper Windows streams, so the behaviour that prompted this PR cannot be reproduced. See https://github.com/pawamoy/duty/issues/23#issuecomment-2501699128 for a summary of the findings.

**Pull request closed without merging**